prov/sockets/src/sock_av.c

```diff
@@ -176,43 +189,43 @@ static int sock_check_table_in(struct sock_av *_av, struct sockaddr_in *addr,
 
 	for (i = 0, ret = 0; i < count; i++) {
 		if (_av->table_hdr->stored == _av->table_hdr->size) {
-			if (_av->table_hdr->req_sz) {
-				if (fi_addr)
-					fi_addr[i] = FI_ADDR_NOTAVAIL;
-				sock_av_report_error(_av, context, i, FI_ENOSPC);
-				SOCK_LOG_ERROR("Cannot insert to AV table\n");
-				continue;
-			} else {
+			index = sock_av_get_next_index(_av);
+			if (index < 0) {
 				new_count = _av->table_hdr->size * 2;
-				table_sz = SOCK_AV_TABLE_SZ(new_count, _av->attr.name);
-				old_sz = SOCK_AV_TABLE_SZ(_av->table_hdr->size, _av->attr.name);
+				table_sz = SOCK_AV_TABLE_SZ(new_count,
+							     _av->attr.name);
+				old_sz = SOCK_AV_TABLE_SZ(_av->table_hdr->size,
+							   _av->attr.name);
 
 				if (_av->attr.name) {
 					new_addr = sock_mremap(_av->table_hdr,
-							       old_sz, table_sz);
+								old_sz, table_sz);
 					if (new_addr == MAP_FAILED) {
 						if (fi_addr)
 							fi_addr[i] = FI_ADDR_NOTAVAIL;
-						sock_av_report_error(_av,
-							context, i, FI_ENOMEM);
+						sock_av_report_error(_av, context,
+								      i, FI_ENOMEM);
 						continue;
 					}
 					_av->idx_arr[_av->table_hdr->stored] = _av->table_hdr->stored;
 				} else {
 					new_addr = realloc(_av->table_hdr,
-								table_sz);
+							    table_sz);
 					if (!new_addr) {
 						if (fi_addr)
 							fi_addr[i] = FI_ADDR_NOTAVAIL;
-						sock_av_report_error(_av,
-							context, i, FI_ENOMEM);
+						sock_av_report_error(_av, context,
+								      i, FI_ENOMEM);
 						continue;
 					}
 				}
 				_av->table_hdr = new_addr;
 				_av->table_hdr->size = new_count;
 				sock_update_av_table(_av, new_count);
+				index = _av->table_hdr->stored++;
 			}
+		} else {
```

Comment:
Yes, we don't reduce the stored count, juts mark them invalid and it's job of the <code>sock_av_get_index</code>to find an invalid index for reuse and we don't increase the <code>stored</code> variable in that case.

>Also, in this code what happens if the stored count is incremented, but then sock_av_is_valid_address fails? It seems like it would invalidly update the stored count.

Good catch. I need to fix that.

> It seems that sock_av_remove should decrement the stored count

Since the address can be removed from any index and <code>stored</code> is used to track next available index until <code>_av->table_hdr->stored < _av_table_hdr->size</code>, we have to iterate over the av table every time we want to insert a new one. The current design allows fast <code>fi_av_insert</code> path when <code># of inserted address < _av_table_hdr->size</code>.